### PR TITLE
Mi-1159 add logic to filter posts with embedded links

### PR DESCRIPTION
**Pull request by hrishavkmr**

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  modules-download-mode: readonly
+
+linters-settings:
+  goconst:
+    min-len: 2
+    min-occurrences: 2
+  gofmt:
+    simplify: true
+  goimports:
+    local-prefixes: github.com/mattermost/mattermost-plugin-autolink
```
Comment (chetanyakan) [inline .golangci.yml]: rename this

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,9 @@
+module github.com/Brightscout/mattermost-plugin-link-filter
+
+go 1.14
+
+require (
+	github.com/mattermost/mattermost-server/v5 v5.23.1
```
Comment (chetanyakan) [inline go.mod:6]: We should either use the minimum server version supported by the APIs or verify the MM version in use.

File: plugin.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      {
+        "key": "RejectPlainLinks",
+        "display_name": "Reject Plain Links:",
+        "type": "bool",
+        "help_text": "If set the plugin will also filter posts containing plain text links in addition to filtering embedded text links."
+      },
+      {
+        "key": "WarningMessage",
+        "display_name": "Warning Message:",
+        "type": "text",
```
Comment (chetanyakan) [inline plugin.json]: Probably makes sense to use `longtext` here instead.

```suggestion
        "type": "longtext",
```

File: plugin.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        "help_text": "If post is rejected, this warning message will be sent to the user. Place `%s` where you want to include the forbidden URL Scheme in the message.",
+        "placeholder": "E.g., Your post has been rejected by the Link Filter, because the following URL Scheme is not allowed: `%s`.",
+        "default": "Your post has been rejected by the Link Filter, because the following URL Scheme is not allowed: `%s`."
+      },
+      {
+        "key": "AllowedProtocolList",
+        "display_name": "Allowed Protocols List:",
+        "type": "longtext",
```
Comment (chetanyakan) [inline plugin.json]: Maybe `text` would be sufficient.
```suggestion
        "type": "text",
```

File: plugin.json [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        "help_text": "The protocols to allow, separated by commas. Capitalization and punctuation insensitive.",
+        "placeholder": "E.g., http, https, mailto, mattermost",
+        "default": "http,https"
```
Comment (chetanyakan) [inline plugin.json]: From the e-mail
> The approved ones we definitely want to allow are http:, https:, mailto:. You also have mattermost: which I assume is required for Mattermost to function. All others will be excluded.

File: server/plugin.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+func (p *Plugin) OnActivate() error {
+	regexString := `\[(?P<text>.*?)\]\((?P<protocol>\w+)://(?P<host>[^\n)]+)?\)`
+	embeddedLinkRegex, err := regexp.Compile(regexString)
+	if err != nil {
+		return err
+	}
+
+	regexString = `(?P<protocol>\w+)://(?P<host>[^\n)]+)?`
```
Comment (chetanyakan) [inline server/plugin.go]: nit - Create constants for these regexes with a comment of what they would match.

File: server/plugin.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	var invalidURLProtocols []string
+	set := make(map[string]bool)
+
+	for _, loc := range detectedURLs {
+		// loc contains the index of relevant groups
+		// [0-1] start and end position of regex
+		// [2-3] start and end position of "text"
+		// [4-5] start and end position of "protocol"
+		// [6-7] start and end position of "host"
+		protocol := string(postText[loc[4]:loc[5]])
```
Comment (chetanyakan) [inline server/plugin.go]: nit - you can also assign `loc[4]`, `loc[5]` to meaningfully named constants for better understandability.

File: server/plugin.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+		// The case when detected url has length 6 i.e., the url is plain link
+		// then the detected url will have no "text" and
+		// start and end position of "protocol" will be 2-3 and not 4-5
+		if len(loc) == 6 {
+			protocol = string(postText[loc[2]:loc[3]])
```
Comment (chetanyakan) [inline server/plugin.go]: same here

File: server/util/util.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package util
+
+import (
+	"strings"
+)
+
+func TrimString(textList []string) []string {
```
Comment (chetanyakan) [inline server/util/util.go:7]: Is this used anywhere?

**Review by chetanyakan — CHANGES_REQUESTED** [review]

File: server/util/util.go (hunk 9 above)
Comment (hrishavkmr) [reply server/util/util.go:7]: yes, https://github.com/Brightscout/mattermost-plugin-link-filter/pull/1/files#diff-065f09c83ba4f3083feef6a8ea1f19ab382d993d8cb51045b63a329c07c51ea0R104
this function is used to trim space and remove the empty strings from the list.

File: plugin.json (hunk 5 above)
Comment (hrishavkmr) [reply plugin.json]: done

File: server/plugin.go (hunk 8 above)
Comment (hrishavkmr) [reply server/plugin.go]: done

File: server/plugin.go (hunk 7 above)
Comment (hrishavkmr) [reply server/plugin.go]: done

File: server/plugin.go (hunk 6 above)
Comment (hrishavkmr) [reply server/plugin.go]: done

File: .golangci.yml (hunk 1 above)
Comment (hrishavkmr) [reply .golangci.yml]: done

File: .golangci.yml [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  modules-download-mode: readonly
+
+linters-settings:
+  goconst:
+    min-len: 2
+    min-occurrences: 2
+  gofmt:
+    simplify: true
+  goimports:
+    local-prefixes: github.com/mattermost/mattermost-plugin-link-filter
```
Comment (chetanyakan) [inline .golangci.yml]: Should be same as the value in `go.mod`
```suggestion
    local-prefixes: github.com/Brightscout/mattermost-plugin-link-filter
```

**Review by chetanyakan — APPROVED** [review]

**Pull request merged**